src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -198,86 +88,204 @@ object LoadMatrix {
+                off = newoff + 1
+                missing = true
+                0L
+              } else {
+                fatal(s"Error parsing matrix. Invalid Int64 at column: $colNum, row: ${ if (rowID == null) "null" else rowID.toString() } in file: $file")
+              }
+            } else if (line.length == newoff || line(newoff) == sep) {
+              off = newoff + 1
+              if (isNegative) -v else v
+            } else {
+              fatal(s"Error parsing matrix. Invalid Int64 at column: $colNum, row: ${ if (rowID == null) "null" else rowID.toString() } in file: $file")
+            }
+          }
+
+          def getFloat(file: String, rowID: Annotation, colNum: Int): Float = {
+            var newoff = line.indexOf(sep, off)
+            if (newoff == -1)
+              newoff = line.length
+            val v = line.substring(off, newoff)
+            off = newoff + 1
+            if (v == missingValue) {
+              missing = true
+              0.0F
+            } else {
+              try {
+              v.toFloat
+              } catch {
+                case _: NumberFormatException => fatal(s"Error parsing matrix: $v is not a Float32. column: $colNum, row: ${ if (rowID == null) "null" else rowID.toString() } in file: $file")
+              }
+            }
+          }
+
+          def getDouble(file: String, rowID: Annotation, colNum: Int): Double = {
+            var newoff = line.indexOf(sep, off)
+            if (newoff == -1)
+              newoff = line.length
+            val v = line.substring(off, newoff)
+            off = newoff + 1
+            if (v == missingValue) {
+              missing = true
+              0.0
+            } else {
+              try {
+                v.toDouble
+              } catch {
+                case _: NumberFormatException => fatal(s"Error parsing matrix: $v is not a Float64. column: $colNum, row: ${ if (rowID == null) "null" else rowID.toString() } in file: $file")
+              }
+            }
+          }
+
+          var ii = 0
+          while (ii < nAnnotations) {
+            val t = at.fieldType(ii)
+            ec.set(ii, t match {
+              case _: TInt32 => getInt(fileByPartition(i), null, ii)
+              case _: TInt64 => getLong(fileByPartition(i), null, ii)
+              case _: TFloat32 => getFloat(fileByPartition(i), null, ii)
+              case _: TFloat64 => getDouble(fileByPartition(i), null, ii)
+              case _: TString => getString(fileByPartition(i), null, ii)
+            })
+            ii += 1
+          }
+
+          val rowKey = f()
 
           region.clear()
           rvb.start(matrixType.rowType)
           rvb.startStruct()
-
-          rvb.addString(rowKey)
-          rvb.addString(rowKey)
+          rvb.addAnnotation(t, rowKey)
+          rvb.addAnnotation(t, rowKey)
           rvb.startStruct()
+          ii = 0
+          while (ii < at.size) {
+            rvb.addAnnotation(at.fieldType(ii), ec.a(ii))
```

Comment:
Hmm, reading `ec.a`.  That feels weird, but I guess is OK since the evaluator is functional.

Reply:
yeah, it felt kind of weird to me as well but also maybe marginally better than figuring out somewhere else to store them.